File: 23/2.c
#include <stdio.h>
#include <stdlib.h>

int dest(int p1, int p2, int p3, int current, int size) {
  while (1) {
    current = (current - 1) % size;
    if (current == 0) current = size;
    if (current != p1 && current != p2 && current != p3) return current;
  }
}


void p(int* indices, int size) {
  for (int i=1; i<=size; i++) {
    printf("%d, ", indices[i]);
  }
  printf("\n");
}


void main(int argc, char* argv[]) {
  printf("hello, world\n");
  int rounds = atoi(argv[1]);
  int size = atoi(argv[2]);

  printf("%d %d\n", rounds, size);

  int d = dest(1, 2, 3, 4, size);
  printf("%d\n", d);

  int* indices = calloc(size, sizeof(int));
  indices[3] = 0;
  indices[8] = 1;
  indices[9] = 2;
  indices[1] = 3;
  indices[2] = 4;
  indices[5] = 5;
  indices[4] = 6;
  indices[6] = 7;
  indices[7] = 8;
  for (int i=10; i<=size; i++) {
    indices[i-1] = i;
  }
  p(indices, size);

  int head[4];
  head[0] = 3;
  head[1] = 8;
  head[2] = 9;
  head[3] = 1;

  int newhead[4];

  for (int i=0; i<rounds; i++) {
    if (i%1000 == 0) {
      printf("Round: %d\n", i);
    }
    int d = dest(head[1], head[2], head[3], head[0], size);
    int di = indices[d];

    for (int n=1; n<=size; n++) {
      if (n == head[0]) {
        indices[n] = size - 1;
      } else if (n != head[1] && n != head[2] && n != head[3]) {
        if (indices[n] <= di) {
          indices[n] = (indices[n] - 4 + size) % size;
        } else {
          indices[n] = (indices[n] - 1 + size) % size;
        }
      } else {
        indices[n] = (indices[n] + (di-4) + size) % size;
      }

      if (indices[n] < 4) {
        newhead[indices[n]] = n;
      }
    }

    head[0] = newhead[0];
    head[1] = newhead[1];
    head[2] = newhead[2];
    head[3] = newhead[3];
  }


  p(indices, size);

  int one = indices[1];
  for (int c=1; c<=2; c++) {
    for (int i=1; i<=size; i++) {
      if (indices[i] == one + c) {
        printf("%d\n", i);
        break;
      }
    } 
  }
}

File: 23/2.py
#!/usr/bin/env python

import sys

def dest(picked, current, size):
    while True:
        current = (current - 1) % size
        if current == 0:
            current = size
        if current not in picked:
            return current

rounds = int(sys.argv[1])

lines = [l.rstrip() for l in sys.stdin.readlines()] 

clockwise = [int(i) for i in lines[0]]

size = int(sys.argv[2])
cupcount = size - len(clockwise)

extracups = range(max(clockwise)+1, max(clockwise)+cupcount+1)

clockwise = clockwise + list(extracups)


for i in range(rounds):
    if not i % 10000:
        print(f"round {i+1}")
    current = clockwise.pop(0)
    clockwise.append(current)
    p = clockwise[:3]
    del clockwise[:3]
    d = dest(p, current, size)

    ci = 0
    while not clockwise[ci] == d:
        ci += 1

    nextstart = clockwise[:ci+1]
    del clockwise[:ci+1]

    clockwise[0:0] = p

    clockwise[0:0] = nextstart
    
one = clockwise.index(1)
out = []
for i in range(one+1, one+3):
    out.append(clockwise[i%len(clockwise)])
print(out)

m = 0
c = 0
for i in range(1, size):
    if clockwise[i-1]+1 == clockwise[i]:
        c += 1
    else:
        m = max(m, c)
        c = 0
print(max(m,c))

